pkg/authn/challenge/challenge.go

```diff
@@ -0,0 +1,118 @@
+package challenge
+
+import (
+	"context"
+	"crypto/rsa"
+	"embed"
+	"encoding/json"
+	"net/http"
+
+	"github.com/bacalhau-project/bacalhau/pkg/authn"
+	"github.com/bacalhau-project/bacalhau/pkg/lib/policy"
+	"github.com/bacalhau-project/bacalhau/pkg/system"
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/samber/lo"
+)
+
+//go:embed *.rego
+var policies embed.FS
+
+type policyData struct {
+	SigningKey jwk.Key `json:"signingKey"`
+	NodeID     string  `json:"nodeId"`
+	ClientID   string  `json:"clientId"`
+}
+
+type response struct {
+	PhraseSignature string `json:"phraseSignature"`
+
+	// A base64 encoded PKCS1 public key, as generated by config.encodePublicKey
+	PublicKey string `json:"publicKey"`
+}
+
+type request struct {
+	InputPhrase []byte `json:"inputPhrase"`
+}
+
+type phraser = func() ([]byte, error)
+
+type challengeAuthenticator struct {
+	authnPolicy *policy.Policy
+	getPhrase   phraser
+	key         jwk.Key
+	nodeID      string
+
+	challenge policy.Query[policyData]
+}
+
+func NewAuthenticator(p *policy.Policy, getPhrase phraser, key *rsa.PrivateKey, nodeID string) authn.Authenticator {
+	return challengeAuthenticator{
+		authnPolicy: p,
+		getPhrase:   getPhrase,
+		key:         lo.Must(jwk.New(key)),
+		nodeID:      nodeID,
+		challenge:   policy.AddQuery[policyData](p, authn.PolicyTokenRule),
+	}
+}
+
+// Authenticate implements Authenticator.
+func (authenticator challengeAuthenticator) Authenticate(req *http.Request) (authn.Authentication, error) {
+	var userInput response
+	err := json.NewDecoder(req.Body).Decode(&userInput)
+	if err != nil {
+		return authn.Authentication{Success: false, Reason: err.Error()}, err
+	}
+
+	inputPhrase, err := authenticator.getPhrase()
+	if err != nil {
+		return authn.Authentication{Success: false, Reason: err.Error()}, err
+	}
+
+	err = system.Verify(inputPhrase, userInput.PhraseSignature, userInput.PublicKey)
+	if err != nil {
+		return authn.Authentication{Success: false, Reason: err.Error()}, nil
+	}
+
+	userKey, err := system.DecodePublicKey(userInput.PublicKey)
+	if err != nil {
+		return authn.Authentication{Success: false, Reason: err.Error()}, err
+	}
+
+	data := policyData{
+		SigningKey: authenticator.key,
+		NodeID:     authenticator.nodeID,
+		ClientID:   system.ConvertToClientID(userKey),
+	}
+
+	result, err := authenticator.challenge(req.Context(), data)
+	if err != nil {
+		return authn.Authentication{Success: false, Reason: err.Error()}, err
+	} else if len(result) != 1 {
+		return authn.Authentication{Success: false, Reason: "signature verified but user credentials rejected"}, nil
+	}
+
+	return authn.Authentication{
+		Success: true,
+		Token:   (result[0].Expressions[0].Value).(string),
```

Comment:
Feels a little bit the policy stuff is leaking through.  NBD but might be nice if we can hide it in a future PR?

Reply:
Ah, I see, you mean leaking through from the query abstraction. Yes.

Reply:
I've made the query abstraction handle extracting the result from the policy, with the assumption that it's one result per query atm.